{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Pandas\n",
    "In this notebook, we'll encounter a very useful package for scientific computing in Python: Pandas. We can think of Pandas as \"numpy with labels\". This package is especially useful for data science and biology for a few reasons:\n",
    "* Great for real-world, heterogeneous data\n",
    "* Similar to Excel spreadsheets (but way faster!)\n",
    "* Smartly deals with missing data\n",
    "\n",
    "### At the end of this notebook, you'll be able to:\n",
    "* Create & manipulate Pandas dataframes\n",
    "* Load a tabular file of genetic data into a Pandas dataframe\n",
    "* Index and subset Pandas dataframes\n",
    "* Generate descriptive statistics for Pandas dataframes\n",
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Importing packages\n",
    "\n",
    "Before we can pandas, we need to import it. The convention is to import `pandas` as `pd`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variable          Type         Data/Info\n",
      "----------------------------------------\n",
      "df_students       DataFrame      FIRST_NAME  LAST_NAME  <...> \\n8               65.0  \n",
      "gene_df           DataFrame                 CA1 field  C<...>20869 rows x 232 columns]\n",
      "gene_df_columns   Index        Index(['CA1 field', 'CA2 <...>ype='object', length=232)\n",
      "last_names        Series       0        Smith\\n1    Leib<...> LAST_NAME, dtype: object\n",
      "pd                module       <module 'pandas' from '/U<...>ages/pandas/__init__.py'>\n",
      "row_index         str          gene_symbol\n",
      "students          dict         n=7\n"
     ]
    }
   ],
   "source": [
    "# Import packages\n",
    "import pandas as pd\n",
    "\n",
    "# Use whos 'magic command' to see available modules\n",
    "%whos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Create and Manipulate Dataframes \n",
    "The two data structures of Pandas are the `Series` and the `DataFrame`. A `Series` is a one-dimensional onject similar to a list. A `DataFrame` can be thought of as a two-dimensional numpy array or a collection of `Series` objects. Series and dataframes can contain multiple different data types such as integers, strings, and floats, similar to an Excel spreadsheet. Pandas also supports `string` lables unlike numpy arrays which only have numeric labels for their rows and columns. For a more in depth explanation, please visit the [Introduction to Data Structures](https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html) section in the Pandas User Guide. \n",
    "\n",
    "You can create a Pandas dataframe by inputting dictionaries into the Pandas function `pd.DataFrame()`, by reading files, or through functions built into the Pandas package. The function [`pd.read_csv()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) reads a comma- or tab-separated file and returns it as a `dataframe`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a data from a dictionary\n",
    "\n",
    "Below, we'll create a dataframe from a dictionary. Since this is quite a bit to fit into one cell, we're also using a trick where we use a pair of open parentheses to tell Python to *ignore* the line breaks here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FIRST_NAME</th>\n",
       "      <th>LAST_NAME</th>\n",
       "      <th>YEAR</th>\n",
       "      <th>HOME_STATE</th>\n",
       "      <th>AGE</th>\n",
       "      <th>CALC_101_FINAL</th>\n",
       "      <th>ENGLISH_101_FINAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Daniel</td>\n",
       "      <td>Smith</td>\n",
       "      <td>1</td>\n",
       "      <td>NY</td>\n",
       "      <td>18</td>\n",
       "      <td>90.0</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ben</td>\n",
       "      <td>Leibstrom</td>\n",
       "      <td>1</td>\n",
       "      <td>NY</td>\n",
       "      <td>19</td>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kavita</td>\n",
       "      <td>Kanabar</td>\n",
       "      <td>1</td>\n",
       "      <td>PA</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linda</td>\n",
       "      <td>Thiel</td>\n",
       "      <td>4</td>\n",
       "      <td>CA</td>\n",
       "      <td>22</td>\n",
       "      <td>60.0</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Omar</td>\n",
       "      <td>Reichel</td>\n",
       "      <td>2</td>\n",
       "      <td>OK</td>\n",
       "      <td>21</td>\n",
       "      <td>70.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Jane</td>\n",
       "      <td>OConner</td>\n",
       "      <td>2</td>\n",
       "      <td>HI</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Felicia</td>\n",
       "      <td>Rao</td>\n",
       "      <td>3</td>\n",
       "      <td>NY</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Rachel</td>\n",
       "      <td>Crock</td>\n",
       "      <td>1</td>\n",
       "      <td>FL</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bob</td>\n",
       "      <td>McDonald</td>\n",
       "      <td>1</td>\n",
       "      <td>FL</td>\n",
       "      <td>18</td>\n",
       "      <td>98.0</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  FIRST_NAME  LAST_NAME  YEAR HOME_STATE  AGE  CALC_101_FINAL  \\\n",
       "0     Daniel      Smith     1         NY   18            90.0   \n",
       "1        Ben  Leibstrom     1         NY   19            80.0   \n",
       "2     Kavita    Kanabar     1         PA   19             NaN   \n",
       "3      Linda      Thiel     4         CA   22            60.0   \n",
       "4       Omar    Reichel     2         OK   21            70.0   \n",
       "5       Jane    OConner     2         HI   19             NaN   \n",
       "6    Felicia        Rao     3         NY   20             NaN   \n",
       "7     Rachel      Crock     1         FL   17             NaN   \n",
       "8        Bob   McDonald     1         FL   18            98.0   \n",
       "\n",
       "   ENGLISH_101_FINAL  \n",
       "0               80.0  \n",
       "1                NaN  \n",
       "2                NaN  \n",
       "3               40.0  \n",
       "4               50.0  \n",
       "5                NaN  \n",
       "6                NaN  \n",
       "7               60.0  \n",
       "8               65.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the dictionary\n",
    "students = (\n",
    "  {'FIRST_NAME': ['Daniel', 'Ben', 'Kavita', 'Linda',\n",
    "                  'Omar','Jane', 'Felicia', 'Rachel',\n",
    "                  'Bob'],\n",
    "   'LAST_NAME': ['Smith', 'Leibstrom', 'Kanabar', 'Thiel',\n",
    "                 'Reichel', 'OConner', 'Rao', 'Crock',\n",
    "                 'McDonald'],\n",
    "   'YEAR': [1, 1, 1, 4, 2, 2, 3, 1, 1],\n",
    "   'HOME_STATE': ['NY', 'NY', 'PA', 'CA', 'OK', 'HI',\n",
    "                  'NY','FL', 'FL'],\n",
    "   'AGE': [18, 19, 19, 22, 21, 19, 20, 17, 18],\n",
    "   'CALC_101_FINAL': [90, 80, None, 60, 70, None, None,\n",
    "                      None, 98],\n",
    "   'ENGLISH_101_FINAL': [80, None, None, 40, 50, None,\n",
    "                         None, 60, 65]} )\n",
    "\n",
    "# Create the dataframe from our dictionary\n",
    "df_students = pd.DataFrame(students)\n",
    "df_students"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can access a list of all of the students' last names using the following notation: `df['column_name']`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        Smith\n",
       "1    Leibstrom\n",
       "2      Kanabar\n",
       "3        Thiel\n",
       "4      Reichel\n",
       "5      OConner\n",
       "6          Rao\n",
       "7        Crock\n",
       "8     McDonald\n",
       "Name: LAST_NAME, dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_students['LAST_NAME'] # Access LAST_NAME column"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One column of the dataframe is a **series**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.series.Series"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_names = df_students['LAST_NAME']\n",
    "type(last_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading from a file example\n",
    "Below we will create a dataframe by reading the file `brainarea_vs_genes_exp_w_reannotations.tsv` which contains information on gene expression accross multiple brain areas. we'll use the `read_csv` function to import our gene expression data as a DataFrame.\n",
    "\n",
    ">**About this dataset:**\n",
    "This dataset was created by Derek Howard and Abigail Mayes for the purpose of accelerating advances in data mining of open brain transcriptome data for polygenetic brain disorders. The data comes from normalized microarray datasets of gene expression from 6 adult human brains that was released by the Allen Brain Institute and then processed into the dataframe we will see below. For more information on this dataset please visit the <a href = \"https://github.com/derekhoward/HBAsets\"> HBAsets repository</a>. \n",
    "\n",
    "Below, we'll use the `pandas.read_csv` function to import our gene expression data as a **data frame**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.856487</td>\n",
       "      <td>-1.773695</td>\n",
       "      <td>-0.678679</td>\n",
       "      <td>-0.986914</td>\n",
       "      <td>0.826986</td>\n",
       "      <td>0.948039</td>\n",
       "      <td>0.935427</td>\n",
       "      <td>1.120774</td>\n",
       "      <td>-1.018554</td>\n",
       "      <td>...</td>\n",
       "      <td>0.277830</td>\n",
       "      <td>0.514923</td>\n",
       "      <td>0.733368</td>\n",
       "      <td>-0.104286</td>\n",
       "      <td>-0.910245</td>\n",
       "      <td>1.039610</td>\n",
       "      <td>-0.155167</td>\n",
       "      <td>-0.444398</td>\n",
       "      <td>-0.901361</td>\n",
       "      <td>-0.236790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>0.257664</td>\n",
       "      <td>-1.373085</td>\n",
       "      <td>-0.619923</td>\n",
       "      <td>-0.636275</td>\n",
       "      <td>0.362799</td>\n",
       "      <td>0.353296</td>\n",
       "      <td>0.422766</td>\n",
       "      <td>0.346853</td>\n",
       "      <td>-0.812015</td>\n",
       "      <td>...</td>\n",
       "      <td>1.074116</td>\n",
       "      <td>0.821031</td>\n",
       "      <td>1.219272</td>\n",
       "      <td>0.901213</td>\n",
       "      <td>-1.522431</td>\n",
       "      <td>0.598719</td>\n",
       "      <td>-1.709745</td>\n",
       "      <td>-0.054156</td>\n",
       "      <td>-1.695843</td>\n",
       "      <td>-1.155961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>-0.089614</td>\n",
       "      <td>-0.546903</td>\n",
       "      <td>0.282914</td>\n",
       "      <td>-0.528926</td>\n",
       "      <td>0.507916</td>\n",
       "      <td>0.577696</td>\n",
       "      <td>0.647671</td>\n",
       "      <td>0.306824</td>\n",
       "      <td>0.089958</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030265</td>\n",
       "      <td>-0.187367</td>\n",
       "      <td>-0.428358</td>\n",
       "      <td>-0.465863</td>\n",
       "      <td>-0.136936</td>\n",
       "      <td>1.229487</td>\n",
       "      <td>-0.110680</td>\n",
       "      <td>-0.118175</td>\n",
       "      <td>-0.139776</td>\n",
       "      <td>0.123829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>0.552415</td>\n",
       "      <td>-0.635485</td>\n",
       "      <td>-0.954995</td>\n",
       "      <td>-0.259745</td>\n",
       "      <td>-1.687391</td>\n",
       "      <td>-1.756847</td>\n",
       "      <td>-1.640242</td>\n",
       "      <td>-1.733110</td>\n",
       "      <td>-0.091695</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.058505</td>\n",
       "      <td>0.207109</td>\n",
       "      <td>-0.161808</td>\n",
       "      <td>0.183630</td>\n",
       "      <td>0.948098</td>\n",
       "      <td>-0.977692</td>\n",
       "      <td>0.911896</td>\n",
       "      <td>-0.499357</td>\n",
       "      <td>1.469386</td>\n",
       "      <td>0.557998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>0.758031</td>\n",
       "      <td>1.549857</td>\n",
       "      <td>1.262225</td>\n",
       "      <td>1.338780</td>\n",
       "      <td>-0.289888</td>\n",
       "      <td>-0.407026</td>\n",
       "      <td>-0.358798</td>\n",
       "      <td>-0.589988</td>\n",
       "      <td>0.944684</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.472908</td>\n",
       "      <td>-0.598317</td>\n",
       "      <td>-0.247797</td>\n",
       "      <td>-0.282673</td>\n",
       "      <td>1.396365</td>\n",
       "      <td>0.945043</td>\n",
       "      <td>0.158202</td>\n",
       "      <td>0.572771</td>\n",
       "      <td>0.073088</td>\n",
       "      <td>-0.886780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 233 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol  CA1 field  CA2 field  CA3 field  CA4 field  \\\n",
       "0        A1BG   0.856487  -1.773695  -0.678679  -0.986914   \n",
       "1    A1BG-AS1   0.257664  -1.373085  -0.619923  -0.636275   \n",
       "2        A1CF  -0.089614  -0.546903   0.282914  -0.528926   \n",
       "3         A2M   0.552415  -0.635485  -0.954995  -0.259745   \n",
       "4       A2ML1   0.758031   1.549857   1.262225   1.338780   \n",
       "\n",
       "   Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "0                    0.826986            0.948039   \n",
       "1                    0.362799            0.353296   \n",
       "2                    0.507916            0.577696   \n",
       "3                   -1.687391           -1.756847   \n",
       "4                   -0.289888           -0.407026   \n",
       "\n",
       "   Crus II, lateral hemisphere  Crus II, paravermis  Edinger-Westphal nucleus  \\\n",
       "0                     0.935427             1.120774                 -1.018554   \n",
       "1                     0.422766             0.346853                 -0.812015   \n",
       "2                     0.647671             0.306824                  0.089958   \n",
       "3                    -1.640242            -1.733110                 -0.091695   \n",
       "4                    -0.358798            -0.589988                  0.944684   \n",
       "\n",
       "   ...  temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "0  ...                        0.277830                      0.514923   \n",
       "1  ...                        1.074116                      0.821031   \n",
       "2  ...                       -0.030265                     -0.187367   \n",
       "3  ...                       -0.058505                      0.207109   \n",
       "4  ...                       -0.472908                     -0.598317   \n",
       "\n",
       "   temporal pole, superior aspect  transverse gyri  trochlear nucleus  \\\n",
       "0                        0.733368        -0.104286          -0.910245   \n",
       "1                        1.219272         0.901213          -1.522431   \n",
       "2                       -0.428358        -0.465863          -0.136936   \n",
       "3                       -0.161808         0.183630           0.948098   \n",
       "4                       -0.247797        -0.282673           1.396365   \n",
       "\n",
       "   tuberomammillary nucleus  ventral tegmental area  \\\n",
       "0                  1.039610               -0.155167   \n",
       "1                  0.598719               -1.709745   \n",
       "2                  1.229487               -0.110680   \n",
       "3                 -0.977692                0.911896   \n",
       "4                  0.945043                0.158202   \n",
       "\n",
       "   ventromedial hypothalamic nucleus  vestibular nuclei  zona incerta  \n",
       "0                          -0.444398          -0.901361     -0.236790  \n",
       "1                          -0.054156          -1.695843     -1.155961  \n",
       "2                          -0.118175          -0.139776      0.123829  \n",
       "3                          -0.499357           1.469386      0.557998  \n",
       "4                           0.572771           0.073088     -0.886780  \n",
       "\n",
       "[5 rows x 233 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the list of lists as a data frame\n",
    "gene_df = pd.read_csv('Data/brainarea_vs_genes_exp_w_reannotations.tsv',sep='\\t')\n",
    "\n",
    "#Show the first five rows\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At the moment, the first column  of information above, called the **index**, just contains a list of numbers. We can reassign the row labels by using the method `set_index()`. We can choose any column in our present dataframe to be the row values. Let's assign the row lables to be the `gene_symbol` and reassign the dataframe. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>Heschl's gyrus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.856487</td>\n",
       "      <td>-1.773695</td>\n",
       "      <td>-0.678679</td>\n",
       "      <td>-0.986914</td>\n",
       "      <td>0.826986</td>\n",
       "      <td>0.948039</td>\n",
       "      <td>0.935427</td>\n",
       "      <td>1.120774</td>\n",
       "      <td>-1.018554</td>\n",
       "      <td>0.170282</td>\n",
       "      <td>...</td>\n",
       "      <td>0.277830</td>\n",
       "      <td>0.514923</td>\n",
       "      <td>0.733368</td>\n",
       "      <td>-0.104286</td>\n",
       "      <td>-0.910245</td>\n",
       "      <td>1.039610</td>\n",
       "      <td>-0.155167</td>\n",
       "      <td>-0.444398</td>\n",
       "      <td>-0.901361</td>\n",
       "      <td>-0.236790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.257664</td>\n",
       "      <td>-1.373085</td>\n",
       "      <td>-0.619923</td>\n",
       "      <td>-0.636275</td>\n",
       "      <td>0.362799</td>\n",
       "      <td>0.353296</td>\n",
       "      <td>0.422766</td>\n",
       "      <td>0.346853</td>\n",
       "      <td>-0.812015</td>\n",
       "      <td>0.903358</td>\n",
       "      <td>...</td>\n",
       "      <td>1.074116</td>\n",
       "      <td>0.821031</td>\n",
       "      <td>1.219272</td>\n",
       "      <td>0.901213</td>\n",
       "      <td>-1.522431</td>\n",
       "      <td>0.598719</td>\n",
       "      <td>-1.709745</td>\n",
       "      <td>-0.054156</td>\n",
       "      <td>-1.695843</td>\n",
       "      <td>-1.155961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-0.089614</td>\n",
       "      <td>-0.546903</td>\n",
       "      <td>0.282914</td>\n",
       "      <td>-0.528926</td>\n",
       "      <td>0.507916</td>\n",
       "      <td>0.577696</td>\n",
       "      <td>0.647671</td>\n",
       "      <td>0.306824</td>\n",
       "      <td>0.089958</td>\n",
       "      <td>0.149820</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030265</td>\n",
       "      <td>-0.187367</td>\n",
       "      <td>-0.428358</td>\n",
       "      <td>-0.465863</td>\n",
       "      <td>-0.136936</td>\n",
       "      <td>1.229487</td>\n",
       "      <td>-0.110680</td>\n",
       "      <td>-0.118175</td>\n",
       "      <td>-0.139776</td>\n",
       "      <td>0.123829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.552415</td>\n",
       "      <td>-0.635485</td>\n",
       "      <td>-0.954995</td>\n",
       "      <td>-0.259745</td>\n",
       "      <td>-1.687391</td>\n",
       "      <td>-1.756847</td>\n",
       "      <td>-1.640242</td>\n",
       "      <td>-1.733110</td>\n",
       "      <td>-0.091695</td>\n",
       "      <td>0.003428</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.058505</td>\n",
       "      <td>0.207109</td>\n",
       "      <td>-0.161808</td>\n",
       "      <td>0.183630</td>\n",
       "      <td>0.948098</td>\n",
       "      <td>-0.977692</td>\n",
       "      <td>0.911896</td>\n",
       "      <td>-0.499357</td>\n",
       "      <td>1.469386</td>\n",
       "      <td>0.557998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.758031</td>\n",
       "      <td>1.549857</td>\n",
       "      <td>1.262225</td>\n",
       "      <td>1.338780</td>\n",
       "      <td>-0.289888</td>\n",
       "      <td>-0.407026</td>\n",
       "      <td>-0.358798</td>\n",
       "      <td>-0.589988</td>\n",
       "      <td>0.944684</td>\n",
       "      <td>-0.466327</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.472908</td>\n",
       "      <td>-0.598317</td>\n",
       "      <td>-0.247797</td>\n",
       "      <td>-0.282673</td>\n",
       "      <td>1.396365</td>\n",
       "      <td>0.945043</td>\n",
       "      <td>0.158202</td>\n",
       "      <td>0.572771</td>\n",
       "      <td>0.073088</td>\n",
       "      <td>-0.886780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 232 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             CA1 field  CA2 field  CA3 field  CA4 field  \\\n",
       "gene_symbol                                               \n",
       "A1BG          0.856487  -1.773695  -0.678679  -0.986914   \n",
       "A1BG-AS1      0.257664  -1.373085  -0.619923  -0.636275   \n",
       "A1CF         -0.089614  -0.546903   0.282914  -0.528926   \n",
       "A2M           0.552415  -0.635485  -0.954995  -0.259745   \n",
       "A2ML1         0.758031   1.549857   1.262225   1.338780   \n",
       "\n",
       "             Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "gene_symbol                                                   \n",
       "A1BG                           0.826986            0.948039   \n",
       "A1BG-AS1                       0.362799            0.353296   \n",
       "A1CF                           0.507916            0.577696   \n",
       "A2M                           -1.687391           -1.756847   \n",
       "A2ML1                         -0.289888           -0.407026   \n",
       "\n",
       "             Crus II, lateral hemisphere  Crus II, paravermis  \\\n",
       "gene_symbol                                                     \n",
       "A1BG                            0.935427             1.120774   \n",
       "A1BG-AS1                        0.422766             0.346853   \n",
       "A1CF                            0.647671             0.306824   \n",
       "A2M                            -1.640242            -1.733110   \n",
       "A2ML1                          -0.358798            -0.589988   \n",
       "\n",
       "             Edinger-Westphal nucleus  Heschl's gyrus  ...  \\\n",
       "gene_symbol                                            ...   \n",
       "A1BG                        -1.018554        0.170282  ...   \n",
       "A1BG-AS1                    -0.812015        0.903358  ...   \n",
       "A1CF                         0.089958        0.149820  ...   \n",
       "A2M                         -0.091695        0.003428  ...   \n",
       "A2ML1                        0.944684       -0.466327  ...   \n",
       "\n",
       "             temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "gene_symbol                                                                 \n",
       "A1BG                               0.277830                      0.514923   \n",
       "A1BG-AS1                           1.074116                      0.821031   \n",
       "A1CF                              -0.030265                     -0.187367   \n",
       "A2M                               -0.058505                      0.207109   \n",
       "A2ML1                             -0.472908                     -0.598317   \n",
       "\n",
       "             temporal pole, superior aspect  transverse gyri  \\\n",
       "gene_symbol                                                    \n",
       "A1BG                               0.733368        -0.104286   \n",
       "A1BG-AS1                           1.219272         0.901213   \n",
       "A1CF                              -0.428358        -0.465863   \n",
       "A2M                               -0.161808         0.183630   \n",
       "A2ML1                             -0.247797        -0.282673   \n",
       "\n",
       "             trochlear nucleus  tuberomammillary nucleus  \\\n",
       "gene_symbol                                                \n",
       "A1BG                 -0.910245                  1.039610   \n",
       "A1BG-AS1             -1.522431                  0.598719   \n",
       "A1CF                 -0.136936                  1.229487   \n",
       "A2M                   0.948098                 -0.977692   \n",
       "A2ML1                 1.396365                  0.945043   \n",
       "\n",
       "             ventral tegmental area  ventromedial hypothalamic nucleus  \\\n",
       "gene_symbol                                                              \n",
       "A1BG                      -0.155167                          -0.444398   \n",
       "A1BG-AS1                  -1.709745                          -0.054156   \n",
       "A1CF                      -0.110680                          -0.118175   \n",
       "A2M                        0.911896                          -0.499357   \n",
       "A2ML1                      0.158202                           0.572771   \n",
       "\n",
       "             vestibular nuclei  zona incerta  \n",
       "gene_symbol                                   \n",
       "A1BG                 -0.901361     -0.236790  \n",
       "A1BG-AS1             -1.695843     -1.155961  \n",
       "A1CF                 -0.139776      0.123829  \n",
       "A2M                   1.469386      0.557998  \n",
       "A2ML1                 0.073088     -0.886780  \n",
       "\n",
       "[5 rows x 232 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_index = 'gene_symbol'\n",
    "gene_df = gene_df.set_index(row_index)\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It would help to know what information is in our dataset. In other words, what is across the columns at the top? We can get a list by accessing the `columns` attribute. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CA1 field', 'CA2 field', 'CA3 field', 'CA4 field',\n",
       "       'Crus I, lateral hemisphere', 'Crus I, paravermis',\n",
       "       'Crus II, lateral hemisphere', 'Crus II, paravermis',\n",
       "       'Edinger-Westphal nucleus', 'Heschl's gyrus',\n",
       "       ...\n",
       "       'temporal pole, inferior aspect', 'temporal pole, medial aspect',\n",
       "       'temporal pole, superior aspect', 'transverse gyri',\n",
       "       'trochlear nucleus', 'tuberomammillary nucleus',\n",
       "       'ventral tegmental area', 'ventromedial hypothalamic nucleus',\n",
       "       'vestibular nuclei', 'zona incerta'],\n",
       "      dtype='object', length=232)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Access the columns of our dataframe \n",
    "gene_df_columns = gene_df.columns \n",
    "gene_df_columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indexing Dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Indexing in Pandas works slightly different than in NumPy. Similar to a dictionary, we can index dataframes by their names. \n",
    "\n",
    "The syntax for indexing single locations in a dataframe is `dataframe.loc[row_label,column_label]`. To index an individual column, we use the shorthand syntax `dataframe.[column_label]`. To index an individual row, we use the syntax `dataframe.loc[row_label]`. To index by index #, we use the syntax `dataframe.iloc[index_number]`. Below are some examples on how to access rows, columns, and single values in our dataframe. For more information on indexing dataframes, visit the <a href = \"https://pandas.pydata.org/docs/user_guide/indexing.html#indexing\"> \"Indexing and selecting data\"</a> section in the Pandas User Guide."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene expression values in CA1 field:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "gene_symbol\n",
       "A1BG        0.856487\n",
       "A1BG-AS1    0.257664\n",
       "A1CF       -0.089614\n",
       "A2M         0.552415\n",
       "A2ML1       0.758031\n",
       "              ...   \n",
       "ZYG11A     -0.496398\n",
       "ZYG11B     -0.856866\n",
       "ZYX        -1.941816\n",
       "ZZEF1      -0.015748\n",
       "ZZZ3       -0.924901\n",
       "Name: CA1 field, Length: 20869, dtype: float64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select a single column\n",
    "column = 'CA1 field'\n",
    "print('Gene expression values in CA1 field:')\n",
    "gene_df[column]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene expression of  A1BG  across brain regions:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CA1 field                            0.856487\n",
       "CA2 field                           -1.773695\n",
       "CA3 field                           -0.678679\n",
       "CA4 field                           -0.986914\n",
       "Crus I, lateral hemisphere           0.826986\n",
       "                                       ...   \n",
       "tuberomammillary nucleus             1.039610\n",
       "ventral tegmental area              -0.155167\n",
       "ventromedial hypothalamic nucleus   -0.444398\n",
       "vestibular nuclei                   -0.901361\n",
       "zona incerta                        -0.236790\n",
       "Name: A1BG, Length: 232, dtype: float64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select a single row\n",
    "row = 'A1BG'\n",
    "print('Gene expression of ', row, ' across brain regions:')\n",
    "gene_df.loc[row]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene expression of A1BG in CA1 field:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.8564873784944677"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select an individual value \n",
    "print('Gene expression of A1BG in CA1 field:')\n",
    "gene_df.loc[row, column]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To select multiple different columns, you can use a `list` of all your columns of interest as so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.856487</td>\n",
       "      <td>-1.773695</td>\n",
       "      <td>-0.678679</td>\n",
       "      <td>-0.986914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.257664</td>\n",
       "      <td>-1.373085</td>\n",
       "      <td>-0.619923</td>\n",
       "      <td>-0.636275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-0.089614</td>\n",
       "      <td>-0.546903</td>\n",
       "      <td>0.282914</td>\n",
       "      <td>-0.528926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.552415</td>\n",
       "      <td>-0.635485</td>\n",
       "      <td>-0.954995</td>\n",
       "      <td>-0.259745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.758031</td>\n",
       "      <td>1.549857</td>\n",
       "      <td>1.262225</td>\n",
       "      <td>1.338780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             CA1 field  CA2 field  CA3 field  CA4 field\n",
       "gene_symbol                                            \n",
       "A1BG          0.856487  -1.773695  -0.678679  -0.986914\n",
       "A1BG-AS1      0.257664  -1.373085  -0.619923  -0.636275\n",
       "A1CF         -0.089614  -0.546903   0.282914  -0.528926\n",
       "A2M           0.552415  -0.635485  -0.954995  -0.259745\n",
       "A2ML1         0.758031   1.549857   1.262225   1.338780"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe w/ only CA field entries \n",
    "CA_field_df = gene_df[['CA1 field', \n",
    "                       'CA2 field', \n",
    "                       'CA3 field', \n",
    "                       'CA4 field']]\n",
    "CA_field_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsetting dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Like NumPy arrays, we can subset our original dataframe to only include data that meets our criteria. Our dataframe has data on multiple different brain areas with many gene expression values. You can filter this dataframe using the following syntax:\n",
    "```\n",
    "new_df = original_df[original_df['Column of Interest'] == 'Desired Value']\n",
    "```\n",
    "In plain english, what this is saying is: save a dataframe from the original dataframe, where the original dataframe values in my Column of Interest are equal to my Desired Value. For more information on subsetting,  visit the <a href = \"https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html\"> \"How do I select a subset of a DataFrame\"</a> section in the Pandas documentation. \n",
    "\n",
    "Below we will demonstrate how to execute this by taking a look at the `CA1 field` column in `gene_df`. We will create a dataframe from `gene_df` that only contains genes that showed a certain level of gene expression. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>Heschl's gyrus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCC12</th>\n",
       "      <td>2.089999</td>\n",
       "      <td>0.684837</td>\n",
       "      <td>0.097313</td>\n",
       "      <td>-0.051411</td>\n",
       "      <td>-1.078900</td>\n",
       "      <td>-0.912071</td>\n",
       "      <td>-1.131497</td>\n",
       "      <td>-0.799075</td>\n",
       "      <td>-0.009423</td>\n",
       "      <td>0.889132</td>\n",
       "      <td>...</td>\n",
       "      <td>1.289340</td>\n",
       "      <td>0.885886</td>\n",
       "      <td>1.271053</td>\n",
       "      <td>0.650801</td>\n",
       "      <td>-0.083413</td>\n",
       "      <td>-0.793237</td>\n",
       "      <td>-0.499512</td>\n",
       "      <td>-0.762330</td>\n",
       "      <td>-0.902496</td>\n",
       "      <td>-0.904421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABHD17C</th>\n",
       "      <td>1.716973</td>\n",
       "      <td>0.601041</td>\n",
       "      <td>1.132500</td>\n",
       "      <td>1.354679</td>\n",
       "      <td>-0.923195</td>\n",
       "      <td>-0.887576</td>\n",
       "      <td>-1.122027</td>\n",
       "      <td>-0.926876</td>\n",
       "      <td>0.339678</td>\n",
       "      <td>0.208174</td>\n",
       "      <td>...</td>\n",
       "      <td>0.530983</td>\n",
       "      <td>0.809101</td>\n",
       "      <td>0.763170</td>\n",
       "      <td>-0.093598</td>\n",
       "      <td>-0.313468</td>\n",
       "      <td>-0.164013</td>\n",
       "      <td>-0.537971</td>\n",
       "      <td>1.169105</td>\n",
       "      <td>-0.663245</td>\n",
       "      <td>-0.377181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABI1</th>\n",
       "      <td>2.051762</td>\n",
       "      <td>2.571777</td>\n",
       "      <td>2.472188</td>\n",
       "      <td>2.261170</td>\n",
       "      <td>0.366138</td>\n",
       "      <td>0.507783</td>\n",
       "      <td>0.449606</td>\n",
       "      <td>0.498424</td>\n",
       "      <td>-1.365242</td>\n",
       "      <td>0.472104</td>\n",
       "      <td>...</td>\n",
       "      <td>0.868471</td>\n",
       "      <td>1.197022</td>\n",
       "      <td>0.914783</td>\n",
       "      <td>0.473846</td>\n",
       "      <td>-2.230209</td>\n",
       "      <td>-0.330684</td>\n",
       "      <td>-1.153189</td>\n",
       "      <td>-0.073650</td>\n",
       "      <td>-1.681602</td>\n",
       "      <td>-1.049258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTB</th>\n",
       "      <td>2.489711</td>\n",
       "      <td>2.806688</td>\n",
       "      <td>2.461655</td>\n",
       "      <td>2.340131</td>\n",
       "      <td>-1.296731</td>\n",
       "      <td>-1.334696</td>\n",
       "      <td>-1.158460</td>\n",
       "      <td>-1.461027</td>\n",
       "      <td>0.088209</td>\n",
       "      <td>0.027236</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.370215</td>\n",
       "      <td>-0.946920</td>\n",
       "      <td>-0.197363</td>\n",
       "      <td>-0.094468</td>\n",
       "      <td>0.926183</td>\n",
       "      <td>-0.034337</td>\n",
       "      <td>0.803164</td>\n",
       "      <td>-0.389238</td>\n",
       "      <td>0.596916</td>\n",
       "      <td>0.110921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTR2</th>\n",
       "      <td>2.655049</td>\n",
       "      <td>2.384445</td>\n",
       "      <td>1.728348</td>\n",
       "      <td>1.413585</td>\n",
       "      <td>-1.377474</td>\n",
       "      <td>-1.207922</td>\n",
       "      <td>-1.496018</td>\n",
       "      <td>-1.359462</td>\n",
       "      <td>0.632764</td>\n",
       "      <td>0.259402</td>\n",
       "      <td>...</td>\n",
       "      <td>0.215683</td>\n",
       "      <td>0.142880</td>\n",
       "      <td>0.080649</td>\n",
       "      <td>0.312218</td>\n",
       "      <td>1.793563</td>\n",
       "      <td>-0.028263</td>\n",
       "      <td>-0.189930</td>\n",
       "      <td>-0.389064</td>\n",
       "      <td>-0.484580</td>\n",
       "      <td>-1.478232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 232 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             CA1 field  CA2 field  CA3 field  CA4 field  \\\n",
       "gene_symbol                                               \n",
       "ABCC12        2.089999   0.684837   0.097313  -0.051411   \n",
       "ABHD17C       1.716973   0.601041   1.132500   1.354679   \n",
       "ABI1          2.051762   2.571777   2.472188   2.261170   \n",
       "ACTB          2.489711   2.806688   2.461655   2.340131   \n",
       "ACTR2         2.655049   2.384445   1.728348   1.413585   \n",
       "\n",
       "             Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "gene_symbol                                                   \n",
       "ABCC12                        -1.078900           -0.912071   \n",
       "ABHD17C                       -0.923195           -0.887576   \n",
       "ABI1                           0.366138            0.507783   \n",
       "ACTB                          -1.296731           -1.334696   \n",
       "ACTR2                         -1.377474           -1.207922   \n",
       "\n",
       "             Crus II, lateral hemisphere  Crus II, paravermis  \\\n",
       "gene_symbol                                                     \n",
       "ABCC12                         -1.131497            -0.799075   \n",
       "ABHD17C                        -1.122027            -0.926876   \n",
       "ABI1                            0.449606             0.498424   \n",
       "ACTB                           -1.158460            -1.461027   \n",
       "ACTR2                          -1.496018            -1.359462   \n",
       "\n",
       "             Edinger-Westphal nucleus  Heschl's gyrus  ...  \\\n",
       "gene_symbol                                            ...   \n",
       "ABCC12                      -0.009423        0.889132  ...   \n",
       "ABHD17C                      0.339678        0.208174  ...   \n",
       "ABI1                        -1.365242        0.472104  ...   \n",
       "ACTB                         0.088209        0.027236  ...   \n",
       "ACTR2                        0.632764        0.259402  ...   \n",
       "\n",
       "             temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "gene_symbol                                                                 \n",
       "ABCC12                             1.289340                      0.885886   \n",
       "ABHD17C                            0.530983                      0.809101   \n",
       "ABI1                               0.868471                      1.197022   \n",
       "ACTB                              -0.370215                     -0.946920   \n",
       "ACTR2                              0.215683                      0.142880   \n",
       "\n",
       "             temporal pole, superior aspect  transverse gyri  \\\n",
       "gene_symbol                                                    \n",
       "ABCC12                             1.271053         0.650801   \n",
       "ABHD17C                            0.763170        -0.093598   \n",
       "ABI1                               0.914783         0.473846   \n",
       "ACTB                              -0.197363        -0.094468   \n",
       "ACTR2                              0.080649         0.312218   \n",
       "\n",
       "             trochlear nucleus  tuberomammillary nucleus  \\\n",
       "gene_symbol                                                \n",
       "ABCC12               -0.083413                 -0.793237   \n",
       "ABHD17C              -0.313468                 -0.164013   \n",
       "ABI1                 -2.230209                 -0.330684   \n",
       "ACTB                  0.926183                 -0.034337   \n",
       "ACTR2                 1.793563                 -0.028263   \n",
       "\n",
       "             ventral tegmental area  ventromedial hypothalamic nucleus  \\\n",
       "gene_symbol                                                              \n",
       "ABCC12                    -0.499512                          -0.762330   \n",
       "ABHD17C                   -0.537971                           1.169105   \n",
       "ABI1                      -1.153189                          -0.073650   \n",
       "ACTB                       0.803164                          -0.389238   \n",
       "ACTR2                     -0.189930                          -0.389064   \n",
       "\n",
       "             vestibular nuclei  zona incerta  \n",
       "gene_symbol                                   \n",
       "ABCC12               -0.902496     -0.904421  \n",
       "ABHD17C              -0.663245     -0.377181  \n",
       "ABI1                 -1.681602     -1.049258  \n",
       "ACTB                  0.596916      0.110921  \n",
       "ACTR2                -0.484580     -1.478232  \n",
       "\n",
       "[5 rows x 232 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a dataframe with only genes that have an expression \n",
    "# value greater than 1.7 in 'CA1 field' \n",
    "desired_column = 'CA1 field'\n",
    "desired_value = 1.7\n",
    "new_gene_df = gene_df[gene_df[desired_column] > desired_value]\n",
    "new_gene_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DataFrame Methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas has many useful methods that you can use on your data, including `describe`, `mean`, and more. To learn more about all the different methods that can be used to manipulate and analyze dataframes, please visit the <a href = \"https://pandas.pydata.org/docs/user_guide/index.html\"> Pandas User Guide </a>. We will demonstrate some of these methods below. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `describe` method returns descriptive statistics of all the columns in our dataframe. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>Heschl's gyrus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.003664</td>\n",
       "      <td>0.017002</td>\n",
       "      <td>0.015315</td>\n",
       "      <td>-0.016633</td>\n",
       "      <td>0.093686</td>\n",
       "      <td>0.088810</td>\n",
       "      <td>0.096118</td>\n",
       "      <td>0.087608</td>\n",
       "      <td>0.047124</td>\n",
       "      <td>-0.042517</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.056050</td>\n",
       "      <td>-0.051731</td>\n",
       "      <td>-0.049905</td>\n",
       "      <td>-0.039528</td>\n",
       "      <td>0.059726</td>\n",
       "      <td>0.014856</td>\n",
       "      <td>0.009535</td>\n",
       "      <td>-0.002853</td>\n",
       "      <td>-0.002527</td>\n",
       "      <td>0.013018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.924456</td>\n",
       "      <td>1.129368</td>\n",
       "      <td>1.078987</td>\n",
       "      <td>0.897192</td>\n",
       "      <td>1.146146</td>\n",
       "      <td>1.118501</td>\n",
       "      <td>1.172986</td>\n",
       "      <td>1.136823</td>\n",
       "      <td>0.973224</td>\n",
       "      <td>0.500526</td>\n",
       "      <td>...</td>\n",
       "      <td>0.567950</td>\n",
       "      <td>0.651098</td>\n",
       "      <td>0.636729</td>\n",
       "      <td>0.494012</td>\n",
       "      <td>1.158916</td>\n",
       "      <td>0.897387</td>\n",
       "      <td>0.686320</td>\n",
       "      <td>0.830602</td>\n",
       "      <td>0.723977</td>\n",
       "      <td>0.725770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-4.076424</td>\n",
       "      <td>-5.923691</td>\n",
       "      <td>-5.994731</td>\n",
       "      <td>-3.971984</td>\n",
       "      <td>-2.739924</td>\n",
       "      <td>-2.662897</td>\n",
       "      <td>-2.908676</td>\n",
       "      <td>-2.864308</td>\n",
       "      <td>-3.671242</td>\n",
       "      <td>-1.666268</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.840486</td>\n",
       "      <td>-2.433961</td>\n",
       "      <td>-2.412614</td>\n",
       "      <td>-1.655962</td>\n",
       "      <td>-6.330275</td>\n",
       "      <td>-3.141490</td>\n",
       "      <td>-1.977225</td>\n",
       "      <td>-3.541112</td>\n",
       "      <td>-2.369304</td>\n",
       "      <td>-2.348784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>-0.570475</td>\n",
       "      <td>-0.644093</td>\n",
       "      <td>-0.631248</td>\n",
       "      <td>-0.573605</td>\n",
       "      <td>-0.802651</td>\n",
       "      <td>-0.778933</td>\n",
       "      <td>-0.824404</td>\n",
       "      <td>-0.794683</td>\n",
       "      <td>-0.651691</td>\n",
       "      <td>-0.414882</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.472481</td>\n",
       "      <td>-0.517562</td>\n",
       "      <td>-0.529458</td>\n",
       "      <td>-0.404416</td>\n",
       "      <td>-0.751329</td>\n",
       "      <td>-0.602619</td>\n",
       "      <td>-0.497988</td>\n",
       "      <td>-0.542164</td>\n",
       "      <td>-0.532177</td>\n",
       "      <td>-0.493921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>-0.025821</td>\n",
       "      <td>0.011189</td>\n",
       "      <td>0.006140</td>\n",
       "      <td>-0.044566</td>\n",
       "      <td>0.100558</td>\n",
       "      <td>0.098665</td>\n",
       "      <td>0.109358</td>\n",
       "      <td>0.096585</td>\n",
       "      <td>0.022760</td>\n",
       "      <td>-0.057495</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.091029</td>\n",
       "      <td>-0.083971</td>\n",
       "      <td>-0.089572</td>\n",
       "      <td>-0.051758</td>\n",
       "      <td>0.037410</td>\n",
       "      <td>-0.030474</td>\n",
       "      <td>-0.023799</td>\n",
       "      <td>-0.023471</td>\n",
       "      <td>-0.045003</td>\n",
       "      <td>-0.031623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.561571</td>\n",
       "      <td>0.706946</td>\n",
       "      <td>0.706398</td>\n",
       "      <td>0.547903</td>\n",
       "      <td>0.985159</td>\n",
       "      <td>0.951919</td>\n",
       "      <td>1.007791</td>\n",
       "      <td>0.960813</td>\n",
       "      <td>0.736614</td>\n",
       "      <td>0.312739</td>\n",
       "      <td>...</td>\n",
       "      <td>0.352796</td>\n",
       "      <td>0.412321</td>\n",
       "      <td>0.403382</td>\n",
       "      <td>0.310551</td>\n",
       "      <td>0.844723</td>\n",
       "      <td>0.587382</td>\n",
       "      <td>0.503613</td>\n",
       "      <td>0.508410</td>\n",
       "      <td>0.501191</td>\n",
       "      <td>0.492262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>7.062717</td>\n",
       "      <td>7.387742</td>\n",
       "      <td>6.413603</td>\n",
       "      <td>7.178692</td>\n",
       "      <td>2.679149</td>\n",
       "      <td>2.717237</td>\n",
       "      <td>2.963899</td>\n",
       "      <td>2.857205</td>\n",
       "      <td>7.552440</td>\n",
       "      <td>2.234669</td>\n",
       "      <td>...</td>\n",
       "      <td>2.199291</td>\n",
       "      <td>2.631498</td>\n",
       "      <td>3.065735</td>\n",
       "      <td>2.238555</td>\n",
       "      <td>6.892682</td>\n",
       "      <td>5.968364</td>\n",
       "      <td>7.267837</td>\n",
       "      <td>6.650673</td>\n",
       "      <td>2.723777</td>\n",
       "      <td>2.845665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 232 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CA1 field     CA2 field     CA3 field     CA4 field  \\\n",
       "count  20869.000000  20869.000000  20869.000000  20869.000000   \n",
       "mean       0.003664      0.017002      0.015315     -0.016633   \n",
       "std        0.924456      1.129368      1.078987      0.897192   \n",
       "min       -4.076424     -5.923691     -5.994731     -3.971984   \n",
       "25%       -0.570475     -0.644093     -0.631248     -0.573605   \n",
       "50%       -0.025821      0.011189      0.006140     -0.044566   \n",
       "75%        0.561571      0.706946      0.706398      0.547903   \n",
       "max        7.062717      7.387742      6.413603      7.178692   \n",
       "\n",
       "       Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "count                20869.000000        20869.000000   \n",
       "mean                     0.093686            0.088810   \n",
       "std                      1.146146            1.118501   \n",
       "min                     -2.739924           -2.662897   \n",
       "25%                     -0.802651           -0.778933   \n",
       "50%                      0.100558            0.098665   \n",
       "75%                      0.985159            0.951919   \n",
       "max                      2.679149            2.717237   \n",
       "\n",
       "       Crus II, lateral hemisphere  Crus II, paravermis  \\\n",
       "count                 20869.000000         20869.000000   \n",
       "mean                      0.096118             0.087608   \n",
       "std                       1.172986             1.136823   \n",
       "min                      -2.908676            -2.864308   \n",
       "25%                      -0.824404            -0.794683   \n",
       "50%                       0.109358             0.096585   \n",
       "75%                       1.007791             0.960813   \n",
       "max                       2.963899             2.857205   \n",
       "\n",
       "       Edinger-Westphal nucleus  Heschl's gyrus  ...  \\\n",
       "count              20869.000000    20869.000000  ...   \n",
       "mean                   0.047124       -0.042517  ...   \n",
       "std                    0.973224        0.500526  ...   \n",
       "min                   -3.671242       -1.666268  ...   \n",
       "25%                   -0.651691       -0.414882  ...   \n",
       "50%                    0.022760       -0.057495  ...   \n",
       "75%                    0.736614        0.312739  ...   \n",
       "max                    7.552440        2.234669  ...   \n",
       "\n",
       "       temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "count                    20869.000000                  20869.000000   \n",
       "mean                        -0.056050                     -0.051731   \n",
       "std                          0.567950                      0.651098   \n",
       "min                         -1.840486                     -2.433961   \n",
       "25%                         -0.472481                     -0.517562   \n",
       "50%                         -0.091029                     -0.083971   \n",
       "75%                          0.352796                      0.412321   \n",
       "max                          2.199291                      2.631498   \n",
       "\n",
       "       temporal pole, superior aspect  transverse gyri  trochlear nucleus  \\\n",
       "count                    20869.000000     20869.000000       20869.000000   \n",
       "mean                        -0.049905        -0.039528           0.059726   \n",
       "std                          0.636729         0.494012           1.158916   \n",
       "min                         -2.412614        -1.655962          -6.330275   \n",
       "25%                         -0.529458        -0.404416          -0.751329   \n",
       "50%                         -0.089572        -0.051758           0.037410   \n",
       "75%                          0.403382         0.310551           0.844723   \n",
       "max                          3.065735         2.238555           6.892682   \n",
       "\n",
       "       tuberomammillary nucleus  ventral tegmental area  \\\n",
       "count              20869.000000            20869.000000   \n",
       "mean                   0.014856                0.009535   \n",
       "std                    0.897387                0.686320   \n",
       "min                   -3.141490               -1.977225   \n",
       "25%                   -0.602619               -0.497988   \n",
       "50%                   -0.030474               -0.023799   \n",
       "75%                    0.587382                0.503613   \n",
       "max                    5.968364                7.267837   \n",
       "\n",
       "       ventromedial hypothalamic nucleus  vestibular nuclei  zona incerta  \n",
       "count                       20869.000000       20869.000000  20869.000000  \n",
       "mean                           -0.002853          -0.002527      0.013018  \n",
       "std                             0.830602           0.723977      0.725770  \n",
       "min                            -3.541112          -2.369304     -2.348784  \n",
       "25%                            -0.542164          -0.532177     -0.493921  \n",
       "50%                            -0.023471          -0.045003     -0.031623  \n",
       "75%                             0.508410           0.501191      0.492262  \n",
       "max                             6.650673           2.723777      2.845665  \n",
       "\n",
       "[8 rows x 232 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `mean` and `std` method return the mean and standard deviation of each column in the dataframe, respectfully. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CA1 field                            0.003664\n",
       "CA2 field                            0.017002\n",
       "CA3 field                            0.015315\n",
       "CA4 field                           -0.016633\n",
       "Crus I, lateral hemisphere           0.093686\n",
       "                                       ...   \n",
       "tuberomammillary nucleus             0.014856\n",
       "ventral tegmental area               0.009535\n",
       "ventromedial hypothalamic nucleus   -0.002853\n",
       "vestibular nuclei                   -0.002527\n",
       "zona incerta                         0.013018\n",
       "Length: 232, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CA1 field                            0.924456\n",
       "CA2 field                            1.129368\n",
       "CA3 field                            1.078987\n",
       "CA4 field                            0.897192\n",
       "Crus I, lateral hemisphere           1.146146\n",
       "                                       ...   \n",
       "tuberomammillary nucleus             0.897387\n",
       "ventral tegmental area               0.686320\n",
       "ventromedial hypothalamic nucleus    0.830602\n",
       "vestibular nuclei                    0.723977\n",
       "zona incerta                         0.725770\n",
       "Length: 232, dtype: float64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.std()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging & joining dataframes\n",
    "\n",
    "Let's say we have two different dataframes and we would like to combine the two into one single dataframe. We can use either the `merge` or `join` Pandas methods in order to pull all of this data into one dataframe. \n",
    "\n",
    "![](http://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png)\n",
    "\n",
    "There are different types of joins/merges you can do in Pandas, illustrated <a href=\"http://www.datasciencemadesimple.com/join-merge-data-frames-pandas-python/\">above</a>. Here, we want to do an **inner** merge, where we're only keeping entries with indices that are in both dataframes. We could do this merge based on columns, alternatively.\n",
    "\n",
    "**Inner** is the default kind of join, so we do not need to specify it. And by default, join will use the 'left' dataframe, in other words, the dataframe that is executing the `join` method.\n",
    "\n",
    "If you need more information, look at the <a href=\"https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html\">join</a> and <a href=\"https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html\">merge</a> documentation: you can use either of these to unite your dataframes, though join will be simpler!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is an example of how to join two separate dataframe into one, unified dataframe. We start with one dataframe with only entries from the *temporal pole* and another dataframe with only entries from the CA fields of the hippocampus. We can then join the two dataframes together using the syntax `unified_df = df_1.join(df_2)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.277830</td>\n",
       "      <td>0.514923</td>\n",
       "      <td>0.733368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>1.074116</td>\n",
       "      <td>0.821031</td>\n",
       "      <td>1.219272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-0.030265</td>\n",
       "      <td>-0.187367</td>\n",
       "      <td>-0.428358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-0.058505</td>\n",
       "      <td>0.207109</td>\n",
       "      <td>-0.161808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-0.472908</td>\n",
       "      <td>-0.598317</td>\n",
       "      <td>-0.247797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "gene_symbol                                                                 \n",
       "A1BG                               0.277830                      0.514923   \n",
       "A1BG-AS1                           1.074116                      0.821031   \n",
       "A1CF                              -0.030265                     -0.187367   \n",
       "A2M                               -0.058505                      0.207109   \n",
       "A2ML1                             -0.472908                     -0.598317   \n",
       "\n",
       "             temporal pole, superior aspect  \n",
       "gene_symbol                                  \n",
       "A1BG                               0.733368  \n",
       "A1BG-AS1                           1.219272  \n",
       "A1CF                              -0.428358  \n",
       "A2M                               -0.161808  \n",
       "A2ML1                             -0.247797  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe w/ only Temporal Pole entries \n",
    "temporal_pole_df = gene_df[['temporal pole, inferior aspect', \n",
    "                            'temporal pole, medial aspect', \n",
    "                            'temporal pole, superior aspect']]\n",
    "temporal_pole_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.277830</td>\n",
       "      <td>0.514923</td>\n",
       "      <td>0.733368</td>\n",
       "      <td>0.856487</td>\n",
       "      <td>-1.773695</td>\n",
       "      <td>-0.678679</td>\n",
       "      <td>-0.986914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>1.074116</td>\n",
       "      <td>0.821031</td>\n",
       "      <td>1.219272</td>\n",
       "      <td>0.257664</td>\n",
       "      <td>-1.373085</td>\n",
       "      <td>-0.619923</td>\n",
       "      <td>-0.636275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-0.030265</td>\n",
       "      <td>-0.187367</td>\n",
       "      <td>-0.428358</td>\n",
       "      <td>-0.089614</td>\n",
       "      <td>-0.546903</td>\n",
       "      <td>0.282914</td>\n",
       "      <td>-0.528926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-0.058505</td>\n",
       "      <td>0.207109</td>\n",
       "      <td>-0.161808</td>\n",
       "      <td>0.552415</td>\n",
       "      <td>-0.635485</td>\n",
       "      <td>-0.954995</td>\n",
       "      <td>-0.259745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-0.472908</td>\n",
       "      <td>-0.598317</td>\n",
       "      <td>-0.247797</td>\n",
       "      <td>0.758031</td>\n",
       "      <td>1.549857</td>\n",
       "      <td>1.262225</td>\n",
       "      <td>1.338780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "gene_symbol                                                                 \n",
       "A1BG                               0.277830                      0.514923   \n",
       "A1BG-AS1                           1.074116                      0.821031   \n",
       "A1CF                              -0.030265                     -0.187367   \n",
       "A2M                               -0.058505                      0.207109   \n",
       "A2ML1                             -0.472908                     -0.598317   \n",
       "\n",
       "             temporal pole, superior aspect  CA1 field  CA2 field  CA3 field  \\\n",
       "gene_symbol                                                                    \n",
       "A1BG                               0.733368   0.856487  -1.773695  -0.678679   \n",
       "A1BG-AS1                           1.219272   0.257664  -1.373085  -0.619923   \n",
       "A1CF                              -0.428358  -0.089614  -0.546903   0.282914   \n",
       "A2M                               -0.161808   0.552415  -0.635485  -0.954995   \n",
       "A2ML1                             -0.247797   0.758031   1.549857   1.262225   \n",
       "\n",
       "             CA4 field  \n",
       "gene_symbol             \n",
       "A1BG         -0.986914  \n",
       "A1BG-AS1     -0.636275  \n",
       "A1CF         -0.528926  \n",
       "A2M          -0.259745  \n",
       "A2ML1         1.338780  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Join the two dataframes\n",
    "df_1 = temporal_pole_df\n",
    "df_2 = CA_field_df\n",
    "\n",
    "unified_df = df_1.join(df_2)\n",
    "unified_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Group Task\n",
    "### *Does the number of neurons in the cortex scale with cortex size?*\n",
    "\n",
    "\n",
    "1. Use `pd.read_csv()` to read in our comma-delimited file 'Data/species_brainmass_neurons.csv' and assign it to a dataframe called `brains_df`.\n",
    "2. Separate the dataframe into two dataframes: one that contains all of the primates (`primate_df`; 'Order' is Primata) and one that contains all of the data that are *not* from primates (`nonprimate_df`). **Hint**: You can subset the dataframe by using the syntax `brains_df[brains_df['Order']=='Primata']`.\n",
    "3. Using two separate, subsequent calls to `plt.scatter()` for primates and then non-primates, create a scatterplot with `cortex_mass` on the x axis and `neurons` on the y axis.\n",
    "6. Label your axes.\n",
    "7. (Optional) log scale your axes using `plt.xscale('log')` and `plt.yscale('log')`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Order</th>\n",
       "      <th>cortex_mass_g</th>\n",
       "      <th>Neurons</th>\n",
       "      <th>Other_cells</th>\n",
       "      <th>Neurons_mg</th>\n",
       "      <th>Other_cells_mg</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Microcebus murinus</th>\n",
       "      <td>Primata</td>\n",
       "      <td>0.908</td>\n",
       "      <td>22310400</td>\n",
       "      <td>70649600</td>\n",
       "      <td>24571</td>\n",
       "      <td>77808</td>\n",
       "      <td>Gabi et al., 2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Callithrix jacchus</th>\n",
       "      <td>Primata</td>\n",
       "      <td>5.561</td>\n",
       "      <td>244720000</td>\n",
       "      <td>395340000</td>\n",
       "      <td>44280</td>\n",
       "      <td>71800</td>\n",
       "      <td>Herculano-Houzel et al., 2007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Otolemur garnetti</th>\n",
       "      <td>Primata</td>\n",
       "      <td>6.290</td>\n",
       "      <td>226090000</td>\n",
       "      <td>402070000</td>\n",
       "      <td>37820</td>\n",
       "      <td>63610</td>\n",
       "      <td>Herculano-Houzel et al., 2007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aotus trivirgatus</th>\n",
       "      <td>Primata</td>\n",
       "      <td>10.617</td>\n",
       "      <td>441900000</td>\n",
       "      <td>695420000</td>\n",
       "      <td>41990</td>\n",
       "      <td>65330</td>\n",
       "      <td>Herculano-Houzel et al., 2007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Callimico goeldii</th>\n",
       "      <td>Primata</td>\n",
       "      <td>12.984</td>\n",
       "      <td>357129180</td>\n",
       "      <td>715330820</td>\n",
       "      <td>27505</td>\n",
       "      <td>55093</td>\n",
       "      <td>Gabi et al., 2010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Order  cortex_mass_g    Neurons  Other_cells  \\\n",
       "Species                                                              \n",
       "Microcebus murinus  Primata          0.908   22310400     70649600   \n",
       "Callithrix jacchus  Primata          5.561  244720000    395340000   \n",
       "Otolemur garnetti   Primata          6.290  226090000    402070000   \n",
       "Aotus trivirgatus   Primata         10.617  441900000    695420000   \n",
       "Callimico goeldii   Primata         12.984  357129180    715330820   \n",
       "\n",
       "                    Neurons_mg  Other_cells_mg                         Source  \n",
       "Species                                                                        \n",
       "Microcebus murinus       24571           77808              Gabi et al., 2010  \n",
       "Callithrix jacchus       44280           71800  Herculano-Houzel et al., 2007  \n",
       "Otolemur garnetti        37820           63610  Herculano-Houzel et al., 2007  \n",
       "Aotus trivirgatus        41990           65330  Herculano-Houzel et al., 2007  \n",
       "Callimico goeldii        27505           55093              Gabi et al., 2010  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load in dataset as a dataframe\n",
    "brains_df = pd.read_csv('Data/species_brainmass_neurons.csv',sep=',').set_index('Species')\n",
    "\n",
    "# Separate into two separate dataframes \n",
    "primate_df = brains_df[brains_df['Order']=='Primata']\n",
    "non_primate_df = brains_df[brains_df['Order']!='Primata']\n",
    "\n",
    "# Confirm dataframes\n",
    "primate_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Order</th>\n",
       "      <th>cortex_mass_g</th>\n",
       "      <th>Neurons</th>\n",
       "      <th>Other_cells</th>\n",
       "      <th>Neurons_mg</th>\n",
       "      <th>Other_cells_mg</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sorex fumeus</th>\n",
       "      <td>Eulipotyphla</td>\n",
       "      <td>0.084</td>\n",
       "      <td>9730000</td>\n",
       "      <td>9290000</td>\n",
       "      <td>116727</td>\n",
       "      <td>111754</td>\n",
       "      <td>Sarko et al., 2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mus musculus</th>\n",
       "      <td>Glires</td>\n",
       "      <td>0.173</td>\n",
       "      <td>13688162</td>\n",
       "      <td>12061838</td>\n",
       "      <td>78672</td>\n",
       "      <td>68643</td>\n",
       "      <td>Herculano-Houzel et al., 2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Blarina brevicauda</th>\n",
       "      <td>Eulipotyphla</td>\n",
       "      <td>0.197</td>\n",
       "      <td>11876000</td>\n",
       "      <td>15820000</td>\n",
       "      <td>60214</td>\n",
       "      <td>80729</td>\n",
       "      <td>Sarko et al., 2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Heterocephalus glaber</th>\n",
       "      <td>Glires</td>\n",
       "      <td>0.184</td>\n",
       "      <td>6151875</td>\n",
       "      <td>8398125</td>\n",
       "      <td>33374</td>\n",
       "      <td>45894</td>\n",
       "      <td>Herculano-Houzel et al., 2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Condylura cristata</th>\n",
       "      <td>Eulipotyphla</td>\n",
       "      <td>0.420</td>\n",
       "      <td>17250000</td>\n",
       "      <td>32010000</td>\n",
       "      <td>40777</td>\n",
       "      <td>76995</td>\n",
       "      <td>Sarko et al., 2009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Order  cortex_mass_g   Neurons  Other_cells  \\\n",
       "Species                                                                     \n",
       "Sorex fumeus           Eulipotyphla          0.084   9730000      9290000   \n",
       "Mus musculus                 Glires          0.173  13688162     12061838   \n",
       "Blarina brevicauda     Eulipotyphla          0.197  11876000     15820000   \n",
       "Heterocephalus glaber        Glires          0.184   6151875      8398125   \n",
       "Condylura cristata     Eulipotyphla          0.420  17250000     32010000   \n",
       "\n",
       "                       Neurons_mg  Other_cells_mg  \\\n",
       "Species                                             \n",
       "Sorex fumeus               116727          111754   \n",
       "Mus musculus                78672           68643   \n",
       "Blarina brevicauda          60214           80729   \n",
       "Heterocephalus glaber       33374           45894   \n",
       "Condylura cristata          40777           76995   \n",
       "\n",
       "                                              Source  \n",
       "Species                                               \n",
       "Sorex fumeus                      Sarko et al., 2009  \n",
       "Mus musculus           Herculano-Houzel et al., 2006  \n",
       "Blarina brevicauda                Sarko et al., 2009  \n",
       "Heterocephalus glaber  Herculano-Houzel et al., 2011  \n",
       "Condylura cristata                Sarko et al., 2009  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_primate_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEKCAYAAAD9xUlFAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAhdklEQVR4nO3de5RU5Znv8e9D20B7a6JgRhsdQBgEgTTaoBEZieQATgYhGPEWL6iHcMboEi+RTBxlnBg8wQQFmWExEY0Jo6KjHvESzhKDdyONApKgEzSONBpEjC0meOTynD/qQtNWVVc3e9euXfv3WatXU29V7f30tq2n93t5XnN3REREADpFHYCIiJQPJQUREclSUhARkSwlBRERyVJSEBGRLCUFERHJ2i/qAPZF9+7dvVevXlGHISISK6tWrfrQ3Xvkei7WSaFXr140NjZGHYaISKyY2X/ne07dRyIikqWkICIiWUoKIiKSFesxhVx27NhBU1MTn332WdShSAtdu3alZ8+eVFdXRx2KiBRQcUmhqamJgw46iF69emFmUYcjgLuzdetWmpqa6N27d9ThiOzlkdc2MXvZm7z38XaO6FbDtWP7M3FoXdRhRabiuo8+++wzDj30UCWEMmJmHHroobp7k7LzyGub+P5Dr7Pp4+04sOnj7Xz/odd55LVNUYcWmVgmBTMbb2YLm5ub8z1f4oikLfpvIuVo9rI32b5j115t23fsYvayNyOKKHqxTAruvtTdp9bW1kYdSk5VVVXU19czaNAgzjzzTP7yl7/kfN1JJ50UyPlWrFjBiy++GMixRJLkvY+3t6s9CWKZFMpdTU0Nq1evZt26dXTu3JkFCxbs9fzOnTsBAvsgV1IQ6ZgjutW0qz0JEp8UHnltEyNueZreMx5nxC1PB96XOHLkSDZs2MCKFSsYOXIkp59+OgMHDgTgwAMPBFIf6qeccgoTJkygT58+zJgxg8WLFzN8+HAGDx7MW2+9BcDSpUs54YQTGDp0KF//+tfZvHkz77zzDgsWLGDOnDnU19fz3HPPsWXLFs444wyGDRvGsGHDeOGFFwB45plnqK+vp76+nqFDh7Jt27ZAf1aRuLl2bH9qqqv2aqupruLasf0jiih6FTf7qD0yg0yZPsXMIBMQyOyDnTt38uSTTzJu3DgAXn31VdatW5dzBs6aNWtYv349hxxyCH369OHSSy/llVde4fbbb2fevHncdtttnHzyybz88suYGT/72c/48Y9/zE9+8hOmTZvGgQceyDXXXAPAueeey/Tp0zn55JN59913GTt2LOvXr+fWW29l/vz5jBgxgk8//ZSuXbvu888oEmeZ/881+2iPRCeFQoNM+/JLsX37durr64HUncIll1zCiy++yPDhw/NOyRw2bBiHH344AEcffTRjxowBYPDgwfz6178GUtNtzzrrLN5//30+//zzvMd66qmn+N3vfpd9/Mknn/Dpp58yYsQIrrrqKs477zwmTZpEz549O/wzilSKiUPrEp0EWkt0UghrkCkzptDaAQcckPc9Xbp0yf67U6dO2cedOnXKjkFcfvnlXHXVVZx++umsWLGCmTNn5jzW7t27efnll79wJzBjxgy+8Y1v8MQTTzBixAiWLVvGMccc086fTkQqWaLHFOI2yNTc3ExdXeovmp///OfZ9oMOOmiv8YExY8Ywb9687ONMgnrrrbcYPHgw1113HcOGDeONN94oTeAiEhuJTgpxG2SaOXMmZ555Jscffzzdu3fPto8fP56HH344O9A8d+5cGhsbGTJkCAMHDszOfrrtttsYNGgQQ4YMobq6mtNOOy2qH0VEypS5e9QxdFhDQ4O33k9h/fr1DBgwoOhjaIl76bT3v42IhMPMVrl7Q67nEj2mABpkEhFpKdHdRyIisjclBRERyYplUmirIJ6IiHRMLJNCuRfEExGJq8QPNItI/CR61uDaJbD8JmhugtqeMPoGGDI5sMPH8k6h3JkZV199dfbxrbfemnf1cXvNnDmTurq6bGnuRx99NOfrFixYwD333BPIOX/0ox8FchyRXNpblDLRG+OsXQJLr4DmjYCnvi+9ItUeECWFEHTp0oWHHnqIDz/8MJTjT58+ndWrV/PAAw9w8cUXs3v37r2e37lzJ9OmTeOCCy4I5HxKChKWjnzAJ3pjnOU3wY5WZXh2bE+1B0RJYe0SmDMIZnZLfQ8g4+63335MnTqVOXPmfOG5d955h1NPPZUhQ4YwevRo3n33XQAuuugirrjiCk466ST69OnDgw8+2OZ5BgwYwH777ceHH37IqFGjuPLKK2loaOD2229n5syZ3HrrrQCMGjWK6dOn09DQwIABA1i5ciWTJk2iX79+XH/99dnjTZw4keOPP55jjz2WhQsXAql6SZkCf+eddx4Av/zlLxk+fDj19fV85zvfYdeuXezatYuLLrqIQYMGMXjw4Jw/u0hrHfmAT/TGOM1N7WvvgGQnhRBvxS677DIWL15M6xlSl19+ORdeeCFr167lvPPO44orrsg+9/777/P888/z2GOPMWPGjDbP8Zvf/IZOnTrRo0cPAD7//HMaGxv36rrK6Ny5M42NjUybNo0JEyYwf/581q1bx913383WrVsBWLRoEatWraKxsZG5c+eydetWbrnllmyBv8WLF7N+/Xruv/9+XnjhBVavXk1VVRWLFy9m9erVbNq0iXXr1vH6668zZcqUfbl8khAd+YCPW82yQNXmqWycr70Dkp0UQrwVO/jgg7nggguYO3fuXu0vvfQS5557LgDnn38+zz//fPa5iRMn0qlTJwYOHMjmzZvzHjuzoc4111zD/fffn93/+Kyzzsr7ntNPPx1IleI+9thjOfzww+nSpQt9+vRh48aNAMydO5evfOUrnHjiiWzcuJHf//73XzjO8uXLWbVqFcOGDaO+vp7ly5fz9ttv06dPH95++20uv/xyfvWrX3HwwQcXeaUkyTryAR+3mmWBGn0DVLe6NtU1qfaAJHv2Uci3YldeeSXHHXdc0X81tyyfnalJ9YMf/IDHH38c2FPtdPr06dkNdVoqpjR3y7Lcmcc7d+5kxYoVPPXUU7z00kvsv//+jBo1is8+++wLx3F3LrzwQmbNmvWF59asWcOyZctYsGABS5YsYdGiRUX81JJk147tv9dGV9D2B3yiN8bJzDIKcfZRspNCbc9011GO9gAccsghTJ48mTvvvJOLL74YgJNOOon77ruP888/n8WLFzNy5MiCx7j55pu5+eabA4mnkObmZr70pS+x//7788Ybb/Dyyy9nn6uurmbHjh1UV1czevRoJkyYwPTp0znssMP46KOP2LZtGwcccACdO3fmjDPOoH///nz7298OPWaJv45+wCe6ZtmQyYEmgdaSnRRG35AaQ2jZhRTwrdjVV1/NHXfckX08b948pkyZwuzZs+nRowd33XVXYOfaF+PGjWPBggUMGDCA/v37c+KJJ2afmzp1KkOGDOG4445j8eLF/PCHP2TMmDHs3r2b6upq5s+fT01NDVOmTMnOhMp1JyHlK8p5/4n+gC9DiS+dHfZCENlDpbPLU+u9yiHVhTNr0mB9WFcolc4uJORbMZFyF9Ze5RJPyZ59JCLJnvcvX6CkIJJwiZ73L19QkUkhzuMklUr/TcpX2PP+21vbSKJVcWMKXbt2ZevWrRx66KHZRV0SLXdn69atdO3aNepQJIcw5/23HsTO1DZqeV4pLxU3+2jHjh00NTXlXHgl0enatSs9e/akuro66lCkhEbc8jSbcoxN1HWr4YUZp0YQkUDCZh9VV1fTu3fvqMMQETSIHUdlM6ZgZgPNbImZ/ZuZfSvqeERk32kQO35CTQpmtsjMPjCzda3ax5nZm2a2wcwy5UBPA+a5+/8CgtkIQEQilejidTEV9p3C3cC4lg1mVgXMJ5UEBgLnmNlA4BfA2WY2Gzg05LhEpAQmDq1j1qTB1HWrwUiNJWildHkLdUzB3Z81s16tmocDG9z9bQAzuw+Y4O6zgMvSSeOhfMc0s6nAVICjjjoqlLhFJDiqbRQvUYwp1AEtS5M2AXVm1svMFgL3ALPzvdndF7p7g7s3ZDaXERGRYJTN7CN3f4f0HYCIiEQjijuFTcCRLR73TLeJiEjEokgKK4F+ZtbbzDoDZwOPtucAZjbezBa23v9YRET2TdhTUu8FXgL6m1mTmV3i7juB7wLLgPXAEnf/bXuO6+5L3X1qbW1t8EGLyD5TvaP4Cnv20Tl52p8Angjz3CISDdU7Itabd5XNimYRqQyFNu1JhLVLUtv8Nm8EPPV96RWp9hiIZVLQmIJI+Up8vaPlN+297zukHi+/KZp42imWSUFjCiLlK/H1jpqb2tdeZmKZFESkfCW+3lFtz/a1lxklBREJVOLrHY2+Aapb3RVV16TaY6BsVjS3h5mNB8b37ds36lBEJIdE1zvKzDKK6eyjitt5TURECiu085q6j0REJEtJQUSSae0SmDMIZnZLfY/JOoKwxXJMQURkn2QWmGXWE2QWmEFs+v7DEss7BS1eE5F9EvMFZmGKZVLQ4jUR2ScxX2AWplgmBRGRfRLzBWZhUlIQkeSJ+QKzMCkpiEjyDJkM4+dC7ZGApb6Pn5v4QWbQ7CMRSaohk5UEcojlnYJmH4mIhCOWSUGzjyRptL2llIq6j0TKnLa3lFKK5Z2CSJIkfntLKSklBZEyl/jtLVtTzaJQKSmIlLnEb2/ZUqZmUfNGwPfULFJiCIySgkiZS/z2li2pZlHoYjnQrJ3XJEkyg8mzl73Jex9v54huNVw7tn8yB5lVsyh02nlNJOYeeW1TchLGnEHprqNWao+E6etKH09Maec1kQqVma666ePtOHumq1bsOgbVLAqdkoJIjCVuuqpqFoUulmMKIpKSyOmqqlkUKiUFkRg7olsNm3IkgLKdrrp2SWqmUHNTau+CTLdP6zZ96EdGSUEkxq4d23+vEhhQxtNVc+2L/Mg/gBns+nxPm/ZKjpTGFERibOLQOmZNGkxdtxoMqOtWw6xJg8tz9lGuNQa7d+xJCBladxAp3SmIxNzEoXXlmQRaa89aAq07iEws7xS0n4JIDLVn/2PtlRyZNpOCmR1tZl3S/x5lZleYWbfQIytA+ymIxFCuNQadqqGq895tWncQqWLuFP4T2GVmfYGFwJHAf4QalYhUnlxrDCb+K0yYr3UHZaSYMYXd7r7TzL4JzHP3eWb2WtiBiUgFyrfGQEmgbBSTFHaY2TnAhcD4dFt1eCGJxFOiahBJxSqm+2gK8FXgZnf/g5n1Bn4Rblgi8ZK4GkRSsVQlVSQAI255OufK4iozdrvrzkHKSqEqqW12H5nZCGAm8Nfp1xvg7t4nyCBF4ixfraFd6T+6MncOgBKDlLViuo/uBH4KnAwMAxrS30UkrZhaQxVdvVQqRjFJodndn3T3D9x9a+Yr9MhEYiTXlpm5VHT1UqkIxcw++rWZzQYeAv5fptHdXw0tKpGYab1lZiezbNdRSx2uXpqruqimcUoIikkKJ6S/txyUcODU4MMRia+WNYgys5ECqV6aq7qoKolKSNpMCu7+tVIEIlJJWt857NPso1zVRTOVRJUUJGDFzD6qBW4E/jbd9Axwk7tHVo3OzMYD4/v27RtVCCJtCqx6ab6KoaokKiEoZqB5EbANmJz++gS4K8yg2qKCeJIo+SqGqpKohKCYpHC0u9/o7m+nv/4Z0BoFkVLJVV1UlUQlJMUkhe1mdnLmQXoxm+bViZRKruqiqiQqISlm9tE04J702ALAn0gVxxORUslXXVQkYAWTgplVAee7+1fM7GAAd/+kJJGJiEjJFUwK7r4r03WkZCAiUvmK6T56zcweBR4A/pxpdPeHQotKREQiUUxS6ApsZe8VzE6q7IWIiFSQYlY0TylFICIiEr1iVjTfRerOYC/ufnEoEYmISGSK6T56rMW/uwLfBN4LJxwREYlSMd1H/9nysZndCzwfWkQiIhKZYlY0t9YPOCzoQEREJHrFjClsIzWmYOnvfwSuCzkuERGJQDHdRweVIhAREYlem91HlvJtM/un9OOjzGx4+KGJJNDaJTBnEMzslvq+dknUEUnCFDOm8K/AV4Fz04+3AfNDi0gkqTLbbjZvBHzPtptKDFJCxSSFE9z9MuAzAHf/E9A51KhEkqjQtpsiJVLMOoUd6WqpDmBmPYDdQQdiZkcBc4GPgP9y91uCPodIWdO2m1IGirlTmAs8DBxmZjeTWqPwo2IObmaLzOwDM1vXqn2cmb1pZhvMbEa6eTDwYHql9NDifwSRGMo1dqBtN6UMtJkU3H0x8D1gFvA+MNHdHyjy+HcD41o2pO865gOnAQOBc8xsIPAycImZPQ38qtgfQCR28o0d9BujbTclckUtXnP3N9x9vrvf4e7riz24uz9LqjuopeHAhvR+z58D9wETgCnAje5+KvCNfMc0s6lm1mhmjVu2bCk2FJHykW/s4Pf/V9tuSuTyjim0WLQGexauZd7T2d2LGY/IpQ7Y2OJxE3ACsACYaWbnAu/ke7O7LwQWAjQ0NHyhUJ9I2Ss0dqBtNyVieT/YWy9aM7MDgcuA75AaYwiUu68DvhX0cUXKTm3PdNdRjnaRiBWzeK2bmc0E1gIHAcPc/ep9OOcm4MgWj3um20SSYfQNGjuQspU3KZhZdzObBbwK7ASGuvv17r51H8+5EuhnZr3NrDNwNvBoew5gZuPNbGFzc/M+hiISgSGTNXYgZcvcc3fLm9mfgS3AXaRWMe/F3X/a5sFTZbZHAd2BzaQGku80s78DbgOqgEXufnNHgm9oaPDGxsaOvFVEJLHMbJW7N+R6rtBg8Wz2DC53qCieu5+Tp/0J4ImOHFOk3dYuSc34aW5K9duPvkF/lYvkUWigeWYJ4xAJR2ZNQGYKaGZNACgxiOTQkU12IqcxBSlakPWEVMFUEiCWScHdl7r71Nra2qhDkXIXVD0hVTCVhIhlUhApWlD1hFTBVBKimHUK17f4d5dwwxEJWFBrAlTBVBKi0DqF68zsq+y9yvil8ENqm8YUpGhBrQlQBVNJiELrFCYApwCXAmuAN4AxwBh3f7NkERagdQpSMq1nMUHqjkOLziSGCq1TKNR99DHwj8AGUgvQbk+3zzCzF4MMUAQo79k9WoUsCVFo8dpY4AbgaOCnpGof/dndp5QiMEmYOKwnUAVTSYC8dwru/o/uPppUGetfkCpJ0cPMnjezpSWKT5JCs3tEykIxU1KXuXtjeh+DJnc/mdSGOJHRQHMFCmN2Tzl3R4mUqWK24/xei4cXpds+DCugYmjxWgUKenaPFpuJdEi7Fq+5+5qwApGEC3qPAXVHiXSIVjRLeQh6do8Wm4l0SEf3WRYJXpCze7TlpUiH6E5BKpO2vBTpkFgmBc0+kjZpsZlIh+QtcxEHKnORUNpJTWSfdHQ7TpHyE4eVzyIxFsvuI0kwTTUVCZWSgsSLppqKhEpJQeJF+xqIhEpJQeJFU01FQhXLpKApqQmmqaYiodKUVBGRhOnozmsi4VBJa5GypXUKUlpaZyBS1nSnIKWldQYiZU1JQdpnX7t+tM5ApKwpKUjxgtjNTOsMRMqakoIUL4iuH60zEClrSgpSvCC6frTOQKSsxXL2kZmNB8b37ds36lCSJajdzILcYU1EAhXLOwV3X+ruU2tra6MOJV72dZBYXT8iFS+WdwrSAY9dBY2LgPQK9o6sD8i8LtcGN9r4RqQiKCkkwdoleyeEjMwgcXs+vHN1/WhBmkjFiGX3kbTT8pv4QkLICGJ9gBakiVQMJYUkKPTBH8T6AC1IE6kYSgpJkPeD34IZJNaCNJGKoaSQBLlmDWHQcHEwff6jb2BnVde9mnZWddWsJJEYUlKodJlZQTu2g1Wl2mqPhEkL4e9/GsgpHtk1ghk7LqVpd3d2u9G0uzszdlzKI7tGBHJ8ESkdzT6qZK1nBfmuPesKApwVNHvZm2z6/CQe5KS92l9a9iYTh9YFdh4RCZ/uFCpZiWYFvffx9na1i0j5UlKoZCWaFXREt9bjFYXbRaR8KSlUshLNCrp2bH9qqqv2aqupruLasf0DPY+IhC+WScHMxpvZwubm5qhDKW8lqlU0cWgdsyYNpq5bDQbUdath1qTBGk8QiSFzz7PSNQYaGhq8sbEx6jDKm2oSiUgrZrbK3RtyPafZR5Uqmww2pqeixjf5i0jpKClUolxTUUGF6kSkTbEcU5A25JqKmqFCdSJSgJJCJWpryqkK1YlIHkoKlaitKacqVCcieSgpVKLRNwCW58mAKqOKSEVSUqhEQyaTf7aRa5BZRPJSUgjL2iUwZxDM7Jb6vnZJac9fe2T72kVEUFIIR2ZKaPNGwPdMBS1lYijRamYRqSxKCmEohz2Lh0yG8XPTdwaW+j5+rrqORKQgLV4LQ7nsWTxkspKAiLSL7hTCoD2LRSSmlBTCkK8/v9+YaAefRUTaoO6jMGS6bFpWJ+03Btb8x56xBtUhEpEypKQQltb9+XMG5R98VlIQkTKh7qNSKZfBZxGRApQUSkWDzyISA2WTFMxspJktMLOfmdmLUccTOC0mE5EYCDUpmNkiM/vAzNa1ah9nZm+a2QYzmwHg7s+5+zTgMeDnoQUVVfkJLSYTkRgIe6D5buAO4J5Mg5lVAfOB/wE0ASvN7FF3/136JecCl4QSTesdyUo9A0iLyUSkzIV6p+DuzwIftWoeDmxw97fd/XPgPmACgJkdBTS7+7Z8xzSzqWbWaGaNW7ZsaV9A5VB+QkSkjEUxplAHbGzxuCndBqk7hLsKvdndF7p7g7s39OjRo31n1gwgEZGCymqdgrvfGOoJanumK5fmaBcRkUjuFDYBLYv690y3hU8zgERECooiKawE+plZbzPrDJwNPNqeA5jZeDNb2Nzc3L4zawaQiEhB5p5v28YADm52LzAK6A5sBm509zvN7O+A24AqYJG739yR4zc0NHhjY2NA0YqIJIOZrXL3hlzPhTqm4O7n5Gl/AngizHOLiEj7lc2KZhERiV4sk0KHxxRERKSgWCYFd1/q7lNra2ujDkVEpKLEMimIiEg4lBRERCQrlklBYwoiIuGIZVKI1ZhCVKW6RUQ6oKxqH1WcqEt1i4i0UyzvFGJDpbpFJGaUFMKkUt0iEjOxTAqxGWjOV5JbpbpFpEzFMinEZqBZpbpFJGZimRRiQ6W6RSRmNPsobEMmKwmISGzoTkFERLKUFEREJCuWSSE2s49ERGImlkkhNrOPRERiJpZJQUREwmHuHnUMHWZmW4D/Tj+sBXL1J7Vuz/W61m3dgQ8DCjOXfLEG+d62Xlfo+WKuUTFtYV/HfDEE/d4ormWpfyfzxRX0e3Utg3tfR69lLdDN3XvkfJe7V8QXsLCY9lyvy/GaxihiDfK9bb2u0PPFXKNi2sK+jpV8LUv9O6lrWR7Xsj3v6+i1bOt9ldR9tLTI9lyvy/fesOzL+Yp9b1uvK/R8sddI17Lt5zt63Up9Hff1nLqWwZyzPe/r6LUs+L5Ydx+Fxcwa3b0h6jjiTtcxOLqWwdG1LKyS7hSCtDDqACqErmNwdC2Do2tZgO4UREQkS3cKIiKSpaQgIiJZSgoiIpKlpNAOZtbHzO40swejjiWOzOwAM/u5mf27mZ0XdTxxpt/F4JjZxPTv5P1mNibqeKKWmKRgZovM7AMzW9eqfZyZvWlmG8xsRqFjuPvb7n5JuJHGSzuv6yTgQXf/n8DpJQ+2zLXnWup3sbB2XstH0r+T04Czooi3nCQmKQB3A+NaNphZFTAfOA0YCJxjZgPNbLCZPdbq67DShxwLd1PkdQV6AhvTL9tVwhjj4m6Kv5ZS2N20/1pen34+0RKz85q7P2tmvVo1Dwc2uPvbAGZ2HzDB3WcBf1/iEGOpPdcVaCKVGFaTrD9IitLOa/m7EocXK+25lma2HrgFeNLdXy1tpOUn6f9j1rHnL1dIfWjV5XuxmR1qZguAoWb2/bCDi7F81/Uh4Awz+zeiKT0QRzmvpX4XOyTf7+XlwNeBb5nZtCgCKyeJuVMIgrtvJdXvKB3g7n8GpkQdRyXQ72Jw3H0uMDfqOMpF0u8UNgFHtnjcM90m+0bXNTi6lsHRtSxC0pPCSqCfmfU2s87A2cCjEcdUCXRdg6NrGRxdyyIkJimY2b3AS0B/M2sys0vcfSfwXWAZsB5Y4u6/jTLOuNF1DY6uZXB0LTtOBfFERCQrMXcKIiLSNiUFERHJUlIQEZEsJQUREclSUhARkSwlBRERyVJSkFgzs78ys/vM7C0zW2VmT5jZ37Tj/d3M7B8CiqWXmbmZ/bBFW3cz22FmdwRxjjznfdDM+rTxmu+a2cVhxSCVQ0lBYsvMDHgYWOHuR7v78cD3gS8X+f79gG5AIEkh7Q/AN1o8PhMIbYGUmR0LVGUqfxawiFThN5GClBQkzr4G7HD3BZkGd1/j7s9ZymwzW2dmr5vZWQBmNsrMnjOzR0mVn74FONrMVpvZ7PRrrjWzlWa21sz+Od32TTNbnj7u4Wb2X2b2Vzli+guw3swa0o/PApZknjSz8Wb2GzN7zcyeMrMvp9tPScewOv3cQenzPJtuW2dmI3Oc7zzg/7Q4/iXp2F6x1G5id6Svy1+Ad8xseAevtSSEqqRKnA0CVuV5bhJQD3wF6A6sNLNn088dBwxy9z+ka+4Pcvd6AEttx9iPVO19Ax41s79194fN7AzgMlKbt9zo7n/Mc+77gLPNbDOpzYTeA45IP/c8cKK7u5ldCnwPuBq4BrjM3V8wswOBz4CpwDJ3vzm9Qcz+Oc41Arg3HfsRwD+lf75twNPAmhavbQRGAq/kiVtESUEq1snAve6+C9hsZs8Aw4BPgFfc/Q953jcm/fVa+vGBpJLEs6S6X9YBL7v7vQXO/SvgX4DNwP2tnusJ3G9mhwOdSXU3AbwA/NTMFgMPuXuTma0EFplZNfCIu6/Oca7DgS3pfw8HnnH3jwDM7AGg5fjKB8AxBeIWUfeRxNpvgeM78L4/F3jOgFnuXp/+6uvud6af6wnsBr5sZnn/33H3z0ndwVwNPNjq6XnAHe4+GPgO0DX9nluAS4Ea4AUzO8bdnwX+llR557vN7IIcp9ueOUYRuqZfL5KXkoLE2dNAFzObmmkwsyHpvvfngLPMrMrMepD6cM3VbbINOKjF42XAxekuHMyszswOSw9KLwLOIVVh86o2YvsJcF3mr/YWatlTw//CFnEf7e6vu/v/JlXi+Rgz+2tgs7v/O/AzUt1Cra0H+qb/vRI4xcy+lI73jFav/RtSdzoiean7SGIr3S//TeA2M7uOVD/8O8CVpPruv0qqT92B77n7H83smFbH2GpmL5jZOlJ79F5rZgOAl1KTm/gU+DapXc6ec/fnzWwNqTGKx919fZ7YfkvuWUczgQfM7E+kklrvdPuVZvY1UncivwWeJFXv/1oz25GOI9edwuPAKOApd99kZj8ilfw+At4Amlu8dkT6/CJ5qXS2SIyZWQ3wa2CEu+8yswPd/dP0ncLDwKL0IPlQ4Cp3Pz/SgKXsqftIJMbcfTtwI6kN6AFmmtlqUt1EfwAeSbd3JzUzSaQg3SmIiEiW7hRERCRLSUFERLKUFEREJEtJQUREspQUREQkS0lBRESy/j+fgbhnMXYH3gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Import plotting package\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.scatter(primate_df['cortex_mass_g'],primate_df['Neurons'])\n",
    "plt.scatter(non_primate_df['cortex_mass_g'],non_primate_df['Neurons'])\n",
    "plt.legend(['Primates','Non-Primates'])\n",
    "plt.xlabel('Cortex Mass (g)')\n",
    "plt.ylabel('# Neurons')\n",
    "#plt.xscale('log')\n",
    "#plt.yscale('log')\n",
    "          \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr>\n",
    "\n",
    "## Additional resources\n",
    "See the [Python Data Science Handbook](https://jakevdp.github.io/PythonDataScienceHandbook/03.00-introduction-to-pandas.html) for a more in depth exploration of Pandas, and of course, the [Pandas documentation](https://pandas.pydata.org/docs/user_guide/index.html)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "rise": {
   "theme": "serif"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
